tools/block-generator/generator/generate.go

```diff
@@ -595,71 +494,70 @@ func getAppTxOptions() []interface{} {
+func (g *generator) generateTxGroup(round uint64, intra uint64) ([]txn.SignedTxnWithAD, uint64 /* numTxns */, error) {
 	// TODO: return the number of transactions generated instead of updating intra!!!
 	selection, err := weightedSelection(g.transactionWeights, getTransactionOptions(), paymentTx)
 	if err != nil {
-		return nil, nil, intra, err
+		return nil, 0, err
 	}
 
 	var signedTxns []txn.SignedTxn
-	var ads []txn.ApplyData
-	var nextIntra uint64
+	var numTxns uint64
 	var expectedID uint64
 	switch selection {
 	case paymentTx:
 		var signedTxn txn.SignedTxn
-		var ad txn.ApplyData
-		signedTxn, ad, nextIntra, err = g.generatePaymentTxn(round, intra)
+		signedTxn, numTxns, err = g.generatePaymentTxn(round, intra)
 		signedTxns = []txn.SignedTxn{signedTxn}
-		ads = []txn.ApplyData{ad}
 	case assetTx:
 		var signedTxn txn.SignedTxn
-		var ad txn.ApplyData
-		signedTxn, ad, nextIntra, expectedID, err = g.generateAssetTxn(round, intra)
+		signedTxn, numTxns, expectedID, err = g.generateAssetTxn(round, intra)
 		signedTxns = []txn.SignedTxn{signedTxn}
-		ads = []txn.ApplyData{ad}
 	case applicationTx:
-		signedTxns, ads, nextIntra, expectedID, err = g.generateAppTxn(round, intra)
+		signedTxns, numTxns, expectedID, err = g.generateAppTxn(round, intra)
 	default:
-		return nil, nil, intra, fmt.Errorf("no generator available for %s", selection)
+		return nil, 0, fmt.Errorf("no generator available for %s", selection)
 	}
 
 	if err != nil {
-		return nil, nil, intra, fmt.Errorf("error generating transaction: %w", err)
+		return nil, numTxns, fmt.Errorf("error generating transaction: %w", err)
 	}
 
 	if len(signedTxns) == 0 {
-		return nil, nil, intra, fmt.Errorf("no transactions generated")
+		return nil, numTxns, fmt.Errorf("no transactions generated")
 	}
 
+	txnGroupAD := make([]txn.SignedTxnWithAD, len(signedTxns))
 	for i := range signedTxns {
+		txnGroupAD[i] = txn.SignedTxnWithAD{SignedTxn: signedTxns[i]}
+
+		// for debugging:
 		g.latestPaysetWithExpectedID = append(
 			g.latestPaysetWithExpectedID,
 			txnWithExpectedID{
 				expectedID: expectedID,
 				signedTxn:  &signedTxns[i],
 				intra:      intra,
-				nextIntra:  nextIntra,
+				nextIntra:  intra + numTxns,
 			},
 		)
 	}
-	return signedTxns, ads, nextIntra, nil
+	return txnGroupAD, numTxns, nil
 }
 
 // ---- 1. Pay Transactions ----
 
 // generatePaymentTxn creates a new payment transaction. The sender is always a genesis account, the receiver is random,
 // or a new account.
-func (g *generator) generatePaymentTxn(round uint64, intra uint64) (txn.SignedTxn, txn.ApplyData, uint64 /* nextIntra */, error) {
+func (g *generator) generatePaymentTxn(round uint64, intra uint64) (txn.SignedTxn, uint64 /* numTxns */, error) {
 	selection, err := weightedSelection(g.payTxWeights, getPaymentTxOptions(), paymentPayTx)
 	if err != nil {
-		return txn.SignedTxn{}, txn.ApplyData{}, intra, err
+		return txn.SignedTxn{}, intra, err
```

Comment:
should it return 0 instead of intra since we're now returning numTxns?

Reply:
ec4e5884d6b2caa2b24f7845a1e5a5fe8c74dd92